remotes/docker/registry.go

```diff
@@ -202,11 +209,72 @@ func MatchAllHosts(string) (bool, error) {
 // MatchLocalhost is a host match function which returns true for
 // localhost.
 func MatchLocalhost(host string) (bool, error) {
-	for _, s := range []string{"localhost", "127.0.0.1", "[::1]"} {
-		if len(host) >= len(s) && host[0:len(s)] == s && (len(host) == len(s) || host[len(s)] == ':') {
-			return true, nil
+	switch {
+	case host == "::1":
+		return true, nil
+	case host == "[::1]":
+		return true, nil
+	}
+	h, p, err := net.SplitHostPort(host)
+
+	// addrError helps distinguish between errors of form
+	// "no colon in address" and "too many colons" in address.
+	// The former is fine as the host string need not have a
+	// port. Latter needs to be handled.
+	addrError := &net.AddrError{
+		Err:  "missing port in address",
+		Addr: host,
+	}
+	if err != nil {
+		if err.Error() != addrError.Error() {
+			return false, err
+		} else { // host string without any port specified
+			h = host
 		}
+	} else if len(p) == 0 {
+		return false, errors.New("invalid host name format")
 	}
-	return host == "::1", nil
 
+	// use ipv4 dotted decimal for further checking
+	if h == "::1" || h == "localhost" {
+		h = "127.0.0.1"
+	}
+
+	var ip net.IP
+
+	// handle ip addresses that are in hex format or decimal format.
+	if canBeDottedDecimal := strings.Contains(h, "."); !canBeDottedDecimal {
+		var hexRep string
+		// hex format
```

Comment:
You could use `strconv.PaseInt` to shorten this logic. Also, it should be noted that an IP address can be like "0x7F.010" (127.0.0.8).

However, it turned out that Go's `net.ParseIP` does not support hex and octal form: https://github.com/golang/go/issues/36822#issuecomment-797884240

So it is not necessarily to support these "strange" forms in this PR.

Reply:
Made the changes 👍🏻